File: 17. 电话号码的字母组合/Solution.cpp
#include<bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        unordered_map<char, string> phoneMap{{'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};
        vector<string> res;
        if(digits.empty()) return res;
        string combination;
        backtrack(combination, 0, res, phoneMap, digits);
        return res;
    }

    void backtrack(string combination, int index, vector<string>& res, unordered_map<char, string> phoneMap, string digits){
        if(index == digits.length()){
            res.push_back(combination);
            return ;
        }
        char digit = digits[index];
        string letters = phoneMap.at(digit);
        for(char letter: letters){
            combination.push_back(letter);
            backtrack(combination, index + 1, res, phoneMap, digits);
            combination.pop_back();
        }
    }
};

File: 17. 电话号码的字母组合/Solution.py
class Solution:
    def backtracking(self, start, k, path, ans, digits, num_map):
        if len(path) == k:
            ans.append(path[:])
            return
        for alpha in num_map[digits[start]]:
            path += alpha
            self.backtracking(start+1, k, path, ans, digits, num_map)
            path = path[:-1]

    def letterCombinations(self, digits: str) -> List[str]:
        num_map = {'2':'abc', '3':'def', '4':'ghi', '5':'jkl', '6':'mno', '7':'pqrs', '8':'tuv', '9':'wxyz'}
        path = ''
        ans = []
        if not digits:
            return ans
        self.backtracking(0, len(digits), path, ans, digits, num_map)
        return ans